aws/resource_aws_ecs_cluster_test.go

```diff
@@ -315,6 +351,69 @@ func testAccCheckAWSEcsClusterExists(resourceName string, cluster *ecs.Cluster)
 	}
 }
 
+func testAccCheckAWSEcsClusterCapacityProviders(cluster *ecs.Cluster, expectedProviderNames []string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*AWSClient).ecsconn
+
+		input := &ecs.DescribeClustersInput{
+			Clusters: []*string{cluster.ClusterArn},
+		}
+
+		resp, err := conn.DescribeClusters(input)
+		if err != nil {
+			return fmt.Errorf("error describing ECS Cluster: %s", err)
+		}
+
+		var actual *ecs.Cluster
+		for _, c := range resp.Clusters {
+			if aws.StringValue(c.ClusterArn) == *cluster.ClusterArn {
+				actual = c
+				break
+			}
+
+		}
+		if actual == nil {
+			return fmt.Errorf("ECS Cluster has disappeared")
+		}
+
+		var match bool
+		if length := len(expectedProviderNames); len(actual.CapacityProviders) == length {
+			match = true
+
+			visited := make([]bool, length)
+			for i := 0; i < length; i++ {
+				found := false
+				element := *actual.CapacityProviders[i]
+				for j := 0; j < length; j++ {
+					if visited[j] {
+						continue
+					}
+					if element == expectedProviderNames[j] {
+						visited[j] = true
+						found = true
+						break
+					}
+				}
+				match = match && found
+				if !match {
+					break
+				}
+			}
+		} else {
+			match = false
+		}
+		if !match {
+			actualProviderNames := make([]string, len(actual.CapacityProviders))
+			for i, v := range actual.CapacityProviders {
+				actualProviderNames[i] = *v
+			}
+			return fmt.Errorf("error matching CapacityProviders:\n\texpected %v\n\tgot: %v", expectedProviderNames, actualProviderNames)
+		}
```

Comment:
I believe this could be simplified with `sort.Strings()` and `reflect.DeepEqual()`

```go
actualProviderNames := aws.StringValueSlice(cluster.CapacityProviders)
sort.Strings(actualProviderNames)
sort.Strings(expectedProviderNames)
if !reflect.DeepEqual(actualProviderNames, expectedProviderNames) {
  return fmt.Errorf("error matching CapacityProviders:\n\texpected %v\n\tgot: %v", expectedProviderNames, actualProviderNames)
}
```

Although we should already be performing this type of difference detection testing by nature of refreshing the Terraform state via `d.Set()` and `ImportStateVerify` testing.